### PR TITLE
tests for update_state

### PR DESCRIPTION
**Pull request by LKieser**
unit testing for the update state section of the game loop

### CODE REVIEW

File: tests/code_names.txt
Comment (LKieser) [inline tests/code_names.txt:1]: I needed to add a duplicate of this file to the Tests folder in order for the imported `Game` to access it. If there is a work around you know about to avoid needing to make a duplicate file, I would definitely use it. I just couldn't seem to come up with it myself.

File: tests/update_state_test.py
Comment (LKieser) [inline tests/update_state_test.py:1]: I created a test for almost every possible scenario I could think of in the update state section of the game loop. Let me know if I went a little overboard with the tests. And if so, what type of tests you would see as the more important one and which type of tests are maybe not needed.

Also, I'm not sure why, but for all of these tests, a value that a variable is asserted to be in a previous test seemed to override any of the new values I set for the same variable in a following test. To avoid this, I just ran each test individually in vs code. Did I set up the test wrong somehow? I'll put a comment beside one of the tests where I found this happening.

File: tests/update_state_test.py
Comment (LKieser) [inline tests/update_state_test.py]: This one seemed like a good example of what I mean by the previous value overriding what I try to set the new value as. Here, I set the value of `game.team` to `'blue'` update state will change it to `'red'`. 

When this test is run individually that is what happens and the test passes. 

When I run all the tests together, this test fails as `game.team` is `'blue'` not `'red'`. This could only happen if the original value of `game.team` is `'red'` and so `update_state()` changes it to `'blue'`.  Which would mean that my setting it to `'blue'` at the start of this new test does nothing when I run the tests together. Do you know why this may be happening?

File: tests/code_names.txt
Comment (EricSchrock) [reply tests/code_names.txt:1]: The workaround would be to make the path to the words file an input to the Game and Board classes.

```py
class Board:
    def __init__(self, card_path: str):
        ...

class Game:
    def __init__(self, card_path: str):
        self.board = Board(card_path=card_path)
```

Then you can call `Game(card_path="code_names.txt")` in your main code and `Game(card_path="../code_names.txt")` in your test code.

However, it's not always bad to have separate input files for your tests. You could create test versions of your input, maybe a list of code names with only as many entries as the size of the board so you know exactly what words to expect during your testing. You could have multiple test input files and create a separate `Game` objects in each individual test, initializing each `Game` instance with a different set of cards to trigger certain test cases, instead of reusing the same `Game` object for all tests.

File: tests/update_state_test.py
Comment (EricSchrock) [reply tests/update_state_test.py:1]: I'll give feedback on specific tests as I review them but from just skimming your code it doesn't look like an excessive amount. Over time you'll develop an intuition for the right number and type of tests for different scenarios.

Large numbers of tests can give you higher assurance of correctness but also make it harder to refactor code. Imagine if we had started by adding tests to your code and then did all the data structure and game loop refactoring. It would have taken you longer since you would have had to refactor the tests alongside the code. Now that the API to the state update type functions is fairly set, you can add these tests and use them to check yourself step by step as you change the internals of these functions to use C.

In general, some questions you can ask to decide the right level of testing for a certain project would be...

- How fast am I trying to go?
- How costly are mistakes (minor annoyance, failed class grade, money, human lives, etc.)?
- How quickly can I deploy bug fixes into production (seconds to press a button, days to drive out to a device with a flash drive, months to get everyone to bring their car in for a SW update, etc.)?

File: tests/update_state_test.py
Comment (EricSchrock) [reply tests/update_state_test.py]: This is because you have `game = Game()` at the global scope (outside your functions). If you run the tests individually, `game` has just been initialized, but if you run the tests all together, the state of `game` has been affected by all the tests that come before `test_skip_is_false_so_team_blue_goes_red_and_guess_value_reset`.

To see this in action, temporarily put `print(f"code_number={game.code_number}")` right before your call to `game.update_state()`. Then call `pytest -s` from the `tests` directory. Then add `game = Game()` to the start of the test function and rerun the tests. This test function is using the initialization value of `code_number`, which is `0`, if you run the test individually, but if you run the tests in a group, `test_guessing_is_false_when_guess_value_is_greater_than_one` sets `code_number = 1`.

image
image

Your unit tests should be independent of each other, AKA the changes one test makes should not affect the next test. Each test function should call `game = Game()` to start the test in a known state and you should be able to run all your tests together and get the same results as when you run them individually.

File: tests/update_state_test.py
Comment (EricSchrock) [inline tests/update_state_test.py]: This assumes the function works a certain way. You know internally that this function will skip setting `guess_value` if you set `guess` to a word that can't appear. Since you are using internal knowledge about the function, you are constraining your ability to change the function implementation, even in ways that don't change the final outputs. You could avoid having to do this by combining a few of your tests. Maybe have four tests built around the four possible results of a guess. Then check all three state variables that `check_guess()` sets in each of the four tests. The first test could look like this.

```py
def test_correct_guess():
     game = Game()
     game.code_number = 3
     game.guess = [card.word for card in game.board.cards if card.color == game.team][0]

     assert [card.flipped for card in game.board.cards if card.word == game.guess][0] is False

     game.check_guess()

     assert [card.flipped for card in game.board.cards if card.word == game.guess][0] is True
     assert game.guess_value == 1
     assert game.code_number == 2
```

See if this type of change would make sense in your other groups of tests as well.

**Review by EricSchrock — CHANGES_REQUESTED** [review]
See my responses to your comments and my feedback. This is a good start!

File: tests/update_state_test.py
Comment (EricSchrock) [inline tests/update_state_test.py:17]: Maybe just for this test, add this before this line.

```py
assert game.team == "red"
```

This isn't really to test anything per se, but to document an assumption that your tests are making.

File: tests/update_state_test.py
Comment (EricSchrock) [inline tests/update_state_test.py:43]: For the opponent, neutral, and death word tests, does it matter what value `game.code_number` is set to (you might need to review your `game.py` code to answer this question)? If not, there's no reason to constrain your solution to the way it works right now.

File: tests/update_state_test.py
Comment (EricSchrock) [inline tests/update_state_test.py]: `test_winner_and_game_over_for_death_card` and `test_winner_for_death_card` might be better as `test_red_wins_if_blue_draws_death` and `test_blue_wins_if_red_draws_death`.

File: tests/update_state_test.py
Comment (EricSchrock) [inline tests/update_state_test.py]: Possible clarification.

`'If our last guess was correct and we have guesses left and the game is not over, we should get another guess'`

File: tests/update_state_test.py
Comment (EricSchrock) [inline tests/update_state_test.py]: If the game is over, do you really care what all these values are? You should probably only care about whatever is needed to drive one last proper `game.render()` call. Test that your code works the way it needs to work, not that it works the way it does work. Leave your solution as unconstrained by tests as possible while still proving it works. That way you have flexibility to change your solution without breaking your tests. This will be important as your port from Python to C, as you may have to pick new strategies in C.

File: tests/update_state_test.py
Comment (EricSchrock) [inline tests/update_state_test.py:196]: Add `assert game.guessing is False` before `game.update_state()` to document your assumption.

**Review by EricSchrock — COMMENTED** [review]
Tests are passing and have reasonable coverage of the state update type functions.

1. Review my feedback comments and optionally make changes (your preference).
2. If you make changes, make sure all tests are passing before you merge.

**Review by EricSchrock — APPROVED** [review]
I meant my last comment to be an approval. See the previous comment for feedback to optionally address.

File: tests/update_state_test.py
Comment (LKieser) [reply tests/update_state_test.py:43]: Within `check_guess()`, if the correct answer is guessed, `code_number` is decreased by 1 and if it is anything else, `code_number` should be reset to 0. I think I will leave it the way it is because of this to check if code number is in fact reset.

File: tests/update_state_test.py
Comment (LKieser) [reply tests/update_state_test.py]: That makes sense. I changed it to make sure `game.skip` and  `game.guessing` are true which is what is necessary for `game.render( )` to correctly makes its last render.

**Pull request merged**
